add post on improved error messages (#660)
tidyverse/tidyverse.org: content/blog/tidymodels-errors-q4/index.Rmd

* Improving existing errors: [The outcome went missing](#outcome)
* Do something where we once did nothing: [Predicting with things that can't predict](#predict)
* Make a place and point to it: [Model formulas](#model)

To demonstrate, we'll walk through some examples using the tidymodels packages: 

```{r load-tidymodels}
library(tidymodels)
```

Note that my installed versions include the current dev version of a few tidymodels packages. You can install those versions with:

```{r inst-dev, eval = FALSE}
pak::pak(paste0("tidymodels/", c("tune", "parsnip", "recipes")))
```

> ## The outcome went missing`r emo::ji("ghost")` {#outcome}

Reviewer: ```suggestion
## The outcome went missing `r emo::ji("ghost")` {#outcome}
```

```{r}
mtcars_rec_new <- 
  recipe(mpg ~ ., mtcars) %>%
  step_pca(all_numeric_predictors())

workflow(mtcars_rec_new, linear_spec) %>% fit(mtcars)
```

Works like a charm. That error we saw previously could be much more helpful, though. With the current developmental version of parsnip, this looks like:

```{r no-outcome-parsnip, error = TRUE}
fit(linear_spec, ~ hp, mtcars)
```

Or, with workflows:

```{r no-outcome-workflows, error = TRUE}
workflow(mtcars_rec, linear_spec) %>% fit(mtcars)
```

Much better.

## Predicting with things that can't predict {#predict}

Earlier this year, Dr. Louise E. Sinks put out a [wonderful blog post](https://lsinks.github.io/posts/2023-04-10-tidymodels/tidymodels_tutorial.html) documenting what it felt like to approach the various object types defined in the tidymodels as a newcomer to the collection of packages. They wrote:

> I found it confusing that `fit`, `last_fit`, `fit_resamples`, etc., did not all produce objects that contained the same information and could be acted on by the same functions. 

> This makes sense. While we try to forefront the intended mental model for fitting and predicting with tidymodels in our APIs and documentation, we also needed to be proactive in anticipating common lapses in constructing that mental model.

Reviewer: ```suggestion
This makes sense. While we try to forefront the intended mental model for fitting and predicting with tidymodels in our APIs and documentation, we also need to be proactive in anticipating common challenges in constructing that mental model.
```

#> ! No in-line functions should be used here; use steps to 
#>   define baking actions.
```

But we _just_ used a special with another tidymodels function! Rude!

Or, to demonstrate 2), we pass the preprocessing formula as we ought to but forget to provide the model formula:

```{r, include = FALSE}
gam_wflow <- 
  workflow() %>%
  add_formula(mpg ~ .) %>%
  add_model(gam_spec) 
```

```r
gam_wflow <- 
  workflow() %>%
  add_formula(mpg ~ .) %>%
  add_model(gam_spec) 

gam_wflow %>% fit(mtcars)
#> Error in `fit_xy()`:
#> ! `fit()` must be used with GAM models (due to its use of formulas).
```

Uh, but I _did_ just use `fit()`!

Since the distinction between model formulas and preprocessor formulas comes up in functions across tidymodels, we decide to create a [central page](https://parsnip.tidymodels.org/dev/reference/model_formula.html) that documents the concept itself, hopefully making the syntax associated with it come more easily to users. Then, we link to it _all over the place_. For example, those errors now look like:

```{r, error = TRUE}
recipe(mpg ~ cyl + s(disp, k = 5), mtcars)
```

Or:

```{r, error = TRUE}
gam_wflow %>% fit(mtcars)
```

> While I've only outlined three, there are all sorts of improvements to error messages on their way to the tidymodels packages in upcoming releases. If you happen to stumble across them, we hope they quickly set you back on the right path.`r emo::ji("map")`

Reviewer: ```suggestion
While I've only outlined three, there are all sorts of improvements to error messages on their way to the tidymodels packages in upcoming releases. If you happen to stumble across them, we hope they quickly set you back on the right path. `r emo::ji("map")`
```

---
output: hugodown::hugo_document

slug: tidymodels-errors-q4
title: "Three ways errors are about to get better in tidymodels"

> date: 2023-11-13

Author: Slated with hopes that the linked recipes PR will be merged by Monday, though we can push back if needed.

Author: Recipes PR has been merged! Bumped to this Friday, the 10th.

```{r fit-resamples}
# fit a linear regression model to bootstrap resamples of mtcars
mtcars_res <- fit_resamples(linear_reg(), mpg ~ ., bootstraps(mtcars))

mtcars_res
```

With previous tidymodels versions, mistakenly trying to predict with this object resulted in the following output:

``` r
predict(mtcars_res)
#> Error in UseMethod("predict") : 
#>   no applicable method for 'predict' applied to an object of class
#>   "c('resample_results', 'tune_results', 'tbl_df', 'tbl', 'data.frame')"
```

Some R developers may recognize this error as what results when we didn't define any `predict()` method for `tune_results` objects. We didn't do so because prediction isn't well-defined for tuning results. _But_, this error message does little to help a user understand why that's the case.

We've recently made some changes to error more informatively in this case. We do so by defining a "dummy" `predict()` method for tuning results, implemented only for the sake of erroring more informatively. The same code will now give the following output:

> ```r

Author: In other cases, I've only used `r` in favor of `{r}` for output from older (current CRAN) package versions. I do so here since the inline `{.help}` markup can be difficult to read in markdown output.

```r
recipe(mpg ~ cyl + s(disp, k = 5), mtcars)
#> Error in `inline_check()`:
#> ! No in-line functions should be used here; use steps to 
#>   define baking actions.
```

But we _just_ used a special with another tidymodels function! Rude!

Or, to demonstrate 2), we pass the preprocessing formula as we ought to but forget to provide the model formula:

```{r, include = FALSE}
gam_wflow <- 
  workflow() %>%
  add_formula(mpg ~ .) %>%
  add_model(gam_spec) 
```

```r
gam_wflow <- 
  workflow() %>%
  add_formula(mpg ~ .) %>%
  add_model(gam_spec) 

gam_wflow %>% fit(mtcars)
#> Error in `fit_xy()`:
#> ! `fit()` must be used with GAM models (due to its use of formulas).
```

Uh, but I _did_ just use `fit()`!

Since the distinction between model formulas and preprocessor formulas comes up in functions across tidymodels, we decide to create a [central page](https://parsnip.tidymodels.org/dev/reference/model_formula.html) that documents the concept itself, hopefully making the syntax associated with it come more easily to users. Then, we link to it _all over the place_. For example, those errors now look like:

> <!-- TODO: use the actual error text here -->

Author: Pending https://github.com/tidymodels/recipes/pull/1257.
